### PR TITLE
Added code from the end of session 5

### PR DESCRIPTION
**Pull request by dylanleclair**
Added as much comments as I could without adding more comments then code 🐳 

The "Reuse your code" slides in [session 6](https://docs.google.com/presentation/d/1YNsoRRgh_hrD-Upw5Z175YhwjbWGEVqqgbt4CKfskr4/edit#slide=id.gc77bd79c7d_0_21) will need to be updated to point the player, platform and main to this. 

Please review before merging! 💯

### CODE REVIEW

File: session6/src/main.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+# Initialized the rectangle
+rectangle = pygame.Rect(50, 50, 100, 100)
+
+# Creating the screen and the clock
+screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
+screen.set_alpha(0)  # Make alpha bits transparent
+clock = pygame.time.Clock()
+
+# Create our platforms, passing in x,y,width,height
```
Comment (AlphaRLee) [inline session6/src/main.py:32]: Add space between params (including in comments)
```suggestion
# Create our platforms, passing in x, y, width, height
```

File: session6/src/main.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+direction = [0, 0]
+
+while True:
+    """
+    EVENTS section - how the code reacts when users do things
+    """
+    for event in pygame.event.get():
+        if (
+            event.type == pygame.QUIT
+        ):  # When user clicks the 'x' on the window, close our game
```
Comment (AlphaRLee) [inline session6/src/main.py:60]: Is this your personal convention? Personally don't see how splitting this onto multiple lines makes this any easier to read
```suggestion
        if event.type == pygame.QUIT:  # When user clicks the 'x' on the window, close our game
```

File: session6/src/main.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    if keys_pressed[pygame.K_a]:
+        print("letter pressed")
+    if keys_pressed[pygame.K_4]:
+        print("number pressed")
+
+    # Mouse events
+    mouse_pos = (
+        pygame.mouse.get_pos()
+    )  # Get position of mouse as a tuple representing the
+    # (x, y) coordinate
```
Comment (AlphaRLee) [inline session6/src/main.py:79]: Same as above
```suggestion
    # Get position of mouse as a tuple representing the (x, y) coordinate
    mouse_pos = pygame.mouse.get_pos()
```

File: session6/src/main.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    UPDATE section - manipulate everything on the screen
+    """
+
+    # stop the player if they hit a platform!
+    for platform in platforms:
+        if pygame.sprite.collide_rect(player, platform):
+            player.y_speed = 0  # change their y speed to 0
+            player.rect.y = platform.rect.y - (
+                player.rect.height
+            )  # make sure they aren't "stuck" in a platform
```
Comment (AlphaRLee) [inline session6/src/main.py:97]: Same as above
```suggestion
            player.rect.y = platform.rect.y - player.rect.height  # make sure they aren't "stuck" in a platform
```

File: session6/src/main.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    screen.fill(BLACK)  # Fill the screen with one colour
+
+    # draw the sprite groups
+    players.draw(screen)
+    platforms.draw(screen)
+
+    pygame.display.flip()  # Pygame uses a double-buffer, without this we see half-completed frames
+    clock.tick(
+        FRAME_RATE
+    )  # Pause the clock to always maintain FRAME_RATE frames per second
```
Comment (AlphaRLee) [inline session6/src/main.py:118]: Same as above
```suggestion
    clock.tick(FRAME_RATE)  # Pause the clock to always maintain FRAME_RATE frames per second
```

File: session6/src/platform.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        Args:
+            x: The x coordinate of the platform
+            y: The y coordinate of the platform
+            width: The width of the platform. Must be greater than or equal to the height
+            height: The height of the platform. Recommended to be 50 or less.
+        """
+        super().__init__()
+        self.image = self.create_image(
+            os.path.join("assets", "platform_tile.png"), width, height
+        )
```
Comment (AlphaRLee) [inline session6/src/platform.py:23]: Same as above
```suggestion
        self.image = self.create_image(os.path.join("assets", "platform_tile.png"), width, height)
```

File: session6/src/player.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+
+class Player(pygame.sprite.Sprite):
+    def __init__(self):
+        super().__init__()
+        self.x_speed = 5  # speed in the x direction
+        self.y_speed = 0  # speed in the y direction
+        image_location = os.path.join(
+            "assets", "player.png"
+        )  # find the path of the player image
```
Comment (AlphaRLee) [inline session6/src/player.py:16]: Same as above. Also personally prefer capitalized first letters in comments
```suggestion
        image_location = os.path.join("assets", "player.png")  # Find the path of the player image
```

File: session6/src/main.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+"""
+pygame.init()
+
+# Global constants
+SCREEN_WIDTH = 1000
+SCREEN_HEIGHT = 800
+FRAME_RATE = 60
+
+# Useful colors
+# format is (R,G,B) where each value is 0-255
```
Comment (AlphaRLee) [inline session6/src/main.py:19]: Capitalize first letter of comment and space after commas
```suggestion
# Format is (R, G, B) where each value is 0-255
```

File: session6/src/main.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+BLACK = (0, 0, 0)
+WHITE = (255, 255, 255)
+VIOLET = (139, 7, 54)
```
Comment (AlphaRLee) [inline session6/src/main.py:22]: This seems unused (and pretty niche), do we need to keep it?
```suggestion
```

File: session6/src/main.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+# Initialized the rectangle
+rectangle = pygame.Rect(50, 50, 100, 100)
```
Comment (AlphaRLee) [inline session6/src/main.py:25]: Unused code - safe to delete?
```suggestion
```

File: session6/src/main.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+# Creating the screen and the clock
+screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
+screen.set_alpha(0)  # Make alpha bits transparent
+clock = pygame.time.Clock()
+
+# Create our platforms, passing in x,y,width,height
+platform = Platform(300, 700, 200, 50)
+platform1 = Platform(600, 600, 200, 50)
```
Comment (AlphaRLee) [inline session6/src/main.py:34]: Personal preference - don't name one variable `platform` and the other one `platform1`. Would rather do `platform1` and `platform2`
```suggestion
platform1 = Platform(300, 700, 200, 50)
platform2 = Platform(600, 600, 200, 50)
```

File: session6/src/main.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+# create a sprite group for the platforms
+platforms = pygame.sprite.Group()
+# add the platforms to the sprite group
+platforms.add(platform, platform1)
```
Comment (AlphaRLee) [inline session6/src/main.py:39]: Comment styling nitpick
```suggestion
# Create a sprite group for the platforms
platforms = pygame.sprite.Group()
# Add the platforms to the sprite group
platforms.add(platform, platform1)
```

File: session6/src/main.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+player = Player()
+
+# Create a sprite group for the player
+players = pygame.sprite.Group()
+# Add the player to the sprite group
+player.add(players)
+
+teleport = False
+
+direction = [0, 0]
```
Comment (AlphaRLee) [inline session6/src/main.py:51]: Variable appears unused. Safe to delete?
```suggestion
```

File: session6/src/main.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        if (
+            event.type == pygame.QUIT
+        ):  # When user clicks the 'x' on the window, close our game
+            pygame.quit()
+            sys.exit()
+
+    # Keyboard events
+    keys_pressed = pygame.key.get_pressed()
+
+    direction = [0, 0]
```
Comment (AlphaRLee) [inline session6/src/main.py:67]: Same as above
```suggestion
```

File: session6/src/main.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+    teleport = False
+
+    if keys_pressed[pygame.K_a]:
+        print("letter pressed")
+    if keys_pressed[pygame.K_4]:
+        print("number pressed")
```
Comment (AlphaRLee) [inline session6/src/main.py:73]: 1. The number `4` seems like an arbitrary choice. I'd prefer 1 just to remove the "randomness" of the choice
2. Capitalized first letter in print
```suggestion
    if keys_pressed[pygame.K_a]:
        print("Letter pressed")
    if keys_pressed[pygame.K_1]:
        print("Number pressed")
```

File: session6/src/main.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            event.type == pygame.QUIT
+        ):  # When user clicks the 'x' on the window, close our game
+            pygame.quit()
+            sys.exit()
+
+    # Keyboard events
+    keys_pressed = pygame.key.get_pressed()
+
+    direction = [0, 0]
+    teleport = False
```
Comment (AlphaRLee) [inline session6/src/main.py:68]: This seems out of place inside the key events section. Move this code to the beginning of the `UPDATE section` or to the top of the `while` loop

File: session6/src/main.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    mouse_pos = (
+        pygame.mouse.get_pos()
+    )  # Get position of mouse as a tuple representing the
+    # (x, y) coordinate
+
+    mouse_buttons = pygame.mouse.get_pressed()
+    if mouse_buttons[0]:  # If left mouse pressed
+        teleport = True  # Replace this line
+    if mouse_buttons[2]:  # If right mouse pressed
+        teleport = True  # Replace this line
```
Comment (AlphaRLee) [inline session6/src/main.py:85]: Outdated comments? If students are expected to have the `teleport` function in their players, then update the comments. Else replace `teleport = True` with `pass` instead
```suggestion
    if mouse_buttons[0]:  # If left mouse pressed
        teleport = True
    if mouse_buttons[2]:  # If right mouse pressed
        teleport = True
```

File: session6/src/main.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+
+    """
+    UPDATE section - manipulate everything on the screen
+    """
+
+    # stop the player if they hit a platform!
```
Comment (AlphaRLee) [inline session6/src/main.py:91]: Comment formatting
```suggestion
    # Stop the player if they hit a platform!
```

File: session6/src/main.py [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+    for platform in platforms:
+        if pygame.sprite.collide_rect(player, platform):
+            player.y_speed = 0  # change their y speed to 0
+            player.rect.y = platform.rect.y - (
+                player.rect.height
+            )  # make sure they aren't "stuck" in a platform
+
+    # update the player!
```
Comment (AlphaRLee) [inline session6/src/main.py:99]: Comment formatting. Also this seems to be an overly excited comment?
```suggestion
    # Update the player
```

File: session6/src/main.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+    player.update(keys_pressed)
+
+    # teleport to mouse click
```
Comment (AlphaRLee) [inline session6/src/main.py:102]: Comment formatting
```suggestion
    # Teleport to mouse click
```

File: session6/src/main.py [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    if teleport:
+        player.teleport(mouse_pos)
+
+    """
+    DRAW section - make everything show up on screen
+    """
+    screen.fill(BLACK)  # Fill the screen with one colour
+
+    # draw the sprite groups
```
Comment (AlphaRLee) [inline session6/src/main.py:111]: Comment formatting. Same goes for the rest of this PR
```suggestion
    # Draw the sprite groups
```

File: session6/src/player.py [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+
+GRAVITY = 0.25
+JUMP_VELOCITY = -1 * 12
+
+
+class Player(pygame.sprite.Sprite):
+    def __init__(self):
+        super().__init__()
+        self.x_speed = 5  # speed in the x direction
```
Comment (AlphaRLee) [inline session6/src/player.py:12]: Any reason why this is initialized to 5 not 0?
```suggestion
        self.x_speed = 0  # Speed in the x direction
```

File: session6/src/player.py [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+        self.y_speed = 0  # speed in the y direction
```
Comment (AlphaRLee) [inline session6/src/player.py:13]: Nitpick - define the "common sprite attributes" (image, rect, etc.) at the top of `__init__()` and save these physics-related attributes for the bottom area. This makes copy-pasting easier for sprites (and eventually encourages further abstraction)

File: session6/src/player.py [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+        image_location = os.path.join(
+            "assets", "player.png"
+        )  # find the path of the player image
+        self.image = pygame.image.load(image_location).convert_alpha()  # load the image
+        self.rect = self.image.get_rect()  # get the rect of the image
+        self.rect.x = 400  # have the player start at 400 on x axis
```
Comment (AlphaRLee) [inline session6/src/player.py:19]: Personally prefer setting this as a parameter in the constructor. Also where is the `y` set to? Seems strange to only define `x` and leave `y` to its default
```py
    def __init__(self, x, y):
        ...
        self.rect.x = x
```

File: session6/src/player.py [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+
+        self.jump_cooldown = 0  # the player can jump right away!
+
+    # how do we want to implement the other actions?
+
+    def update(self, keys_pressed):
+
+        # Setup
+
+        print("y speed: " + str(self.y_speed))
```
Comment (AlphaRLee) [inline session6/src/player.py:29]: Necessary to keep here? Useful for debugging and when first introducing concepts, but still needed?
```suggestion
```

File: session6/src/player.py [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+            direction[0] = -1  # x direction goes left
+        if keys_pressed[pygame.K_RIGHT]:
+            direction[0] = 1  # x direction goes right
+
+        if keys_pressed[pygame.K_SPACE]:
+            # this will run if space is pressed
+            # JUMP LOGIC
+            if self.jump_cooldown == 0:  # this will run if jump_cooldown is 0
+                self.jump_cooldown = 30  # reset the cooldown to 30 frames
+                self.y_speed += JUMP_VELOCITY  # add velocity for the jump (must be greater than gravity to work properly)
```
Comment (AlphaRLee) [inline session6/src/player.py:50]: Move all this into a helper function if possible - let's try to keep the `update()` function short

File: session6/src/player.py [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+        # Updates
```
Comment (AlphaRLee) [inline session6/src/player.py:51]: This comment seems out-of-place once you split up the `update()` function with some more helper functions. Maybe delete?
```suggestion
```

File: session6/src/player.py [hunk 28]
```diff
@@ -0,0 +1,72 @@
...
+
+        self.rect.x += direction[0] * self.x_speed
+        # self.rect.y += direction[1] * self.speed
+        # remove gravity, jump code & uncomment the above line to remove platformer physics
```
Comment (AlphaRLee) [inline session6/src/player.py:55]: Move into a helper function called `move`

File: session6/src/player.py [hunk 29]
```diff
@@ -0,0 +1,72 @@
...
+
+        self.y_speed += GRAVITY  # Add gravity to y_speed
+        self.rect.y += self.y_speed  # Add speed to position
+
+        # if jump_cooldown is >= 1, subtract by 1 to lower the cooldown
+        # this happens every frame, so jump_cooldown is
+        # the number of frames between when the player is allowed to jump
+        # when jump_cooldown is 0, we can jump again!
+        if self.jump_cooldown > 0:
+            self.jump_cooldown -= 1
```
Comment (AlphaRLee) [inline session6/src/player.py:65]: Move into a helper function called `fall()`

File: session6/src/main.py (hunk 2 above)
Comment (dylanleclair) [reply session6/src/main.py:60]: Just my autoformatter doing it's thing LOL

File: session6/src/player.py (hunk 22 above)
Comment (dylanleclair) [reply session6/src/player.py:12]: You won't be able to move at all if it's set to zero 😉 x acceleration wasn't covered

File: session6/src/player.py (hunk 24 above)
Comment (dylanleclair) [reply session6/src/player.py:19]: y is initialized when the sprite is. x is in the center so the player falls on a platform on spawn 

i'll add a y to be explicit

File: session6/src/player.py (hunk 26 above)
Comment (dylanleclair) [reply session6/src/player.py:50]: I think it's important to keep there, so no.

File: session6/src/player.py (hunk 27 above)
Comment (dylanleclair) [reply session6/src/player.py:51]: Keeping this. It's to mirror the "structure" of the game loop.

File: session6/src/player.py (hunk 28 above)
Comment (dylanleclair) [reply session6/src/player.py:55]: I'm going to keep it in update to keep the code easy to follow.

File: session6/src/player.py (hunk 29 above)
Comment (dylanleclair) [reply session6/src/player.py:65]: I'm going to keep it in update to keep the code easy to follow. I'll add a comment to separate it from the rest.

File: session6/src/main.py (hunk 16 above)
Comment (dylanleclair) [reply session6/src/main.py:61]: Taken care of (see commits)

File: session6/src/player.py (hunk 24 above)
Comment (dylanleclair) [reply session6/src/player.py:17]: Taken care of! (See commits)

**Comment by dylanleclair** [comment]
Please review one last time!

File: session6/src/player.py [hunk 30]
```diff
@@ -9,30 +9,52 @@
...
+        # EVENTS -------------------------------------
 
-        # Get input
+        # Recall: keys_pressed is passed to us from the main loop
+        # You can check for keys pressed as usual!
+
+
+        # Check if the player needs to teleport
+        if mouse_buttons[0]:  # If left mouse pressed
+            teleport = True  # Replace this line
```
Comment (AlphaRLee) [inline session6/src/player.py:52]: Get rid of outdated comment
```suggestion
            teleport = True
```

File: session6/src/player.py [hunk 31]
```diff
@@ -9,30 +9,52 @@
...
+        if mouse_buttons[2]:  # If right mouse pressed
+            teleport = True  # Replace this line
```
Comment (AlphaRLee) [inline session6/src/player.py:54]: Outdated comment
```suggestion
            teleport = True
```

File: session6/src/player.py [hunk 32]
```diff
@@ -42,29 +64,45 @@ def update(self, keys_pressed):
...
-                self.jump_cooldown = 30  # reset the cooldown to 30 frames
-                self.y_speed += JUMP_VELOCITY  # add velocity for the jump (must be greater than gravity to work properly)
-        # Updates
+            # This will run if SPACE is pressed
+            if self.jump_cooldown == 0:  # This will run if jump_cooldown is 0
+                self.jump_cooldown = 30  # Reset the cooldown to 30 frames
+                self.y_speed += JUMP_VELOCITY  # Add velocity for the jump (must be greater than gravity to work properly)
+        
+        # Set teleport to True if we need to call teleport() in update!
+        if mouse_buttons[0]:  # If left mouse pressed
```
Comment (AlphaRLee) [inline session6/src/player.py:75]: Redundant code for teleporting? Delete one version

File: session6/src/player.py (hunk 31 above)
Comment (AlphaRLee) [reply session6/src/player.py:54]: Resolved in PR #28

File: session6/src/player.py (hunk 32 above)
Comment (AlphaRLee) [reply session6/src/player.py:75]: Resolved in PR #28

File: session6/src/player.py (hunk 30 above)
Comment (AlphaRLee) [reply session6/src/player.py:52]: Resolved in PR #28

**Comment by AlphaRLee** [comment]
Opened PR #28 that merges into this one and makes the minor tweaks mentioned above. @dylanleclair if I don't hear feedback by Monday afternoon, I'll merge it in and merge this PR into main branch

**Pull request merged**